### ISSUE TITLE
Background Job Dashboard

### ISSUE THREAD

**Issue by escowles** [pulibrary/figgy#554]
Use the [Sidekiq API](https://github.com/mperham/sidekiq/wiki/API) to display info about current background jobs.

We should include:
* current number of running jobs
* current number of queued jobs, broken down by default/low queue and type of job.

**Comment by escowles** [comment]
Ideally, users could also find info about in-process items, for example, checking the status of an ingest job by bib ID or path to files being ingested.

**Comment by escowles** [comment]
https://github.com/utgarda/sidekiq-status#tracking-progress-and-storing-data may provide some relevant functionality for proving richer progress info on jobs and a mechanism for tying job and sub-job progress together.

**Comment by hackmastera** [comment]
Use cases: What’s the status of my thing? Is it enqueued? Where is it in the queue? and Did it error?

**Comment by hackmastera** [comment]
Options for implementation:

Option 1: Dashboard on its own page, using the standard sidekiq api. The job knows the id of the object it's operating on, e.g., 
```> default.first.args
=> [{"job_class"=>"CharacterizationJob",
  "job_id"=>"34a6f38f-7a26-4be3-8f8e-df14a4fb6bfd",
  "queue_name"=>"default",
  "priority"=>nil,
  "arguments"=>["1a3434df-547c-476b-afc2-f6e5b9d81de7"],
  "executions"=>0,
  "locale"=>"en"}]
[19] pry(#<CharacterizationJob>)> default.first.args.first['arguments']
=> ["1a3434df-547c-476b-afc2-f6e5b9d81de7"]
```
and that can be used to get titles, parent ids, parent titles, etc. So a user could go the dashboard and ctrl-f to find their object. This page would be refreshed for new data, or use polling. Implementing via ActionCable I think would bring us far enough along the road toward the other option that we would prefer that option.

Option 2: Job information embedded in the FileManager or other record view. This requires the ability to look up a job by the id of the object it operates on. To avoid frequent refresh we would use ActionCable to provide realtime updates. These two requirements suggest a new Job model whose objects are created and destroyed via ActiveJob callbacks on `enqueue`. An ActionCable channel could publish any change in the job's status via a callback on the ActiveRecord model or ChangeSet, depending on whether we want to implement as an active record or a valkyrie object.

Position in queue is harder here; there's not a property of a job (although we could provide the creation timestamp). I'm not sure the best way to achieve that. Probably a separate queue model, but the additional complexity makes me wonder how important that use case is.

Option 3: do both. Provide a link from the filemanager object to the object's position on the full dashboard for people to see where in the queue their item is.

My instinct says that the users would find option 2 much more useful and convenient than option 1, and that position in queue is not nearly as important as "does the job exist and is it still good?" If this is the case it may make sense to move forward on that option and not do a dashboard. However, it will take longer so urgency of need for this feature is something to consider. And some design discussion should probably happen regarding the file manager -- is it important enough to use additional real estate on those little boxes?

**Comment by escowles** [comment]
I agree that option 2 is what our users have asked for, and that some design discussion is called for before we start implementing.  Real estate is at a premium on the file manager (the logical place to show status of the jobs related to the files), so we'll need to be thoughtful about how that shows up.  Maybe it would be possible to provide a summary along the lines of "W files complete, X files generating derivatives, Y files characterizing, Z errors/retries"?

**Comment by hackmastera** [comment]
A summary on the filemanager in lieu of per-object status?

another idea is to put it on the item show page, and also move the 'regenerate' button there. that might relieve some of the design pressure on the file manager.

**Comment by escowles** [comment]
Yes, I think a summary on the show page would be the most useful thing, to avoid having to go into the file manager to check the status.

I'm not sure about the regenerate button though — it seems like you would need the context of each file (and to see if the thumbnail was displaying correctly or not) to do that.

**Comment by hackmastera** [comment]
@escowles sorry, I mean to put the job status / regenerate button on the FileSet show page; corrected above but I initially wrote "item" which could be almost anything 😁 

there is a thumbnail on the fileset show page, right? looking now...

**Comment by hackmastera** [comment]
huh, there's not. Well we could add one.

**Comment by escowles** [comment]
@hackmastera Oh, that makes sense — you'd get a summary on the resource show page, and you could go to the fileset show page for file-level status and the regenerate button could go there too.  And adding the thumbnail there too would be good.

**Comment by hackmastera** [comment]
pending comment from @sdellis I like this plan. I do think this means that "where in the queue is my object?" remains an unmet use case. @escowles how much of a priority is that use case? If it's super important I would like to think a bit more about how it could be implemented.

**Comment by sdellis** [comment]
I like the level of detail in @escowles summary data points, and I agree that position in queue is the least important here. I'm a little fuzzy on what is meant by "filemanager object". If we are going to do anything on the Vue filemanger UI, we can't depend on any Rails views alone. I think this approach is a pragmatic way to fulfill the use cases.

However, my dream is to just have the thumbnails appear in the filemanager gallery pane immediately after upload gets kicked off, with some kind of status as it gets uploaded and the background jobs run... this way you can continue editing things without having to reload the page. The Vue Filemanager would need some way to be notified (either via polling or websockets) each time the status changes.

**Comment by hackmastera** [comment]
@sdellis by 'filemanager object' i meant what you called the "gallery pane" -- sorry for the weird term!

That's a great point that if the thumbnail isn't available yet then there's real estate on the gallery pane. There are other types of jobs. Are the characterization / derivatives jobs really the only one that users care about? If so, this seems pretty do-able.

We'd need to figure out what happens when someone regenerates a derivative, and how to make sure that works on the back end. There would also have to be a well-defined way to retrieve the thumbnail; I assume that's possible already due to iiif.

@sdellis is the question of polling vs websockets immaterial to you? If websockets, is it possible to use ActionCable from the vue app?

**Comment by sdellis** [comment]
Sorry if I wasn't clear, but I think putting this functionality on the Resource show page and the FileSet show pages could probably fill the use cases, but may not be the best experience. 

However, one won't be able to get to the FileSet show page until a thumbnail is somehow viewable in the gallery, right? At the very least, it just needs a fileset id, but is the fileset id only available after the upload is complete?

As an aside, I don't think filemanager real estate is a huge issue... I would rather put functionality where it makes the most sense and figure out how to make it work. I'm not convinced the detail panel is needed if we have the resize option on the gallery... but we'll see.

**Comment by hackmastera** [comment]
@sdellis I'm definitely confused because I think we are saying the same thing / agreeing. Let's talk about it more on stand up.

hilariously, right after I asked about action cable from vue, this came through Ruby Weekly: https://robots.thoughtbot.com/talking-to-actioncable-without-rails

**Comment by hackmastera** [comment]
In the current/old filemanager you have to refresh after upload in order to see new files. However, they do show up (without thumbnails) even if characterization and derivatives have not run.

I think creating a new fileset pane (or whatever you call it) as a response to upload completion is possible (emitting an actioncable message should be easy since you'd just plug it into the existing job) but maybe outside the scope of this ticket. Given a scenario wherein you have a bunch of files, though, which may or may not have thumbnails yet, I think it is a reasonable design choice to replace or overlay the 'default' / no-derivative-yet thumbnail with job status information.

**Comment by hackmastera** [comment]
Summary of discussion on standup:
* We'll continue investigating in the current direction for the rest of the week, then make decisions about feasibility for full implementation.
* We'll observe use in related areas and discuss our design ideas at figgy users meeting next week

Design consensus:
* Brief jobs status overview on a resource show page as outlined by @escowles.
  * This will require the ability to get a job by FileSet id
* Enhancements to file manager as follows:
  * After uploading via browser everything, instead of page refresh, pending uploads will populate the canvas as "fileset boxes" with no thumbnail or title. Title should fill in as the upload completes. Thumbnail should fill in as the derivatives job completes.
  * In each fileset box, if the thumbnail is not yet generated, jobs status info will appear where the thumbnail would be.

Data model requirements:
* A "job tracker" object per-fileset; TBD how vue app will get the id of this object.
  * possibly pass to or receive from browse everything?
  * possibly a property on a fileset object?
  * possibly from a parent-resource pub/sub channel that provides these fileset-level job tracker IDs so they can be subscribed to as they are created? (this is my favorite option so far, seems more loosely coupled)

Javascript-side implementation outline:
* still much TBD here
* But it's clear that a pub/sub channel per fileset, as opposed to per job, is most useful.

**Comment by hackmastera** [comment]
Data model requirements:
* A `JobTrackerBatch` object will be created for each keep track of upload / characterize / derivatives jobs on a per-fileset basis. This object is needed in order to give the front-end an ID before a fileset ID has been created. properties: `id`, `fileset_id` `job_tracker` (one-to-many)
  * vue app will get the id of this object from a parent-level channel that will emit a message each time a new file queued for upload.
    * emit this message as part of the browse everything code
      * Where exactly? at the beginning of `BaseResourceController#browse_everything_files`? this is where the job is started and the page refreshed.
  * `fileset_id` will be filled on success of `BrowseEverythingIngestJob`
* A `JobTracker` will be created for each upload / characterize / derivatives job. These will be updated so they can emit messages via actioncable. they can also be used to present the summary data on a resource show page. properties: `id`, `job_tracker_batch`, `job_id`, `fileset_id`, `job_status`
  * how will this object / these objects stay updated?
    * A background job will start when the application loads, use the sidekiq API to get / parse all jobs into the db, emit message if there's a status change, then wait x seconds and invoke itself.
      * How to ensure this job is not kicked off externally more than once?
      * having it invoke itself ensures it never steps on its own toes
    * Use sidekiq pro and make each job its own batch, then add a callback to update the job tracker object/s and emit messages.
    * Use activejob calbacks? It's unclear whether these are called before or after a job's status has changed, which could be a deal-breaker. otherwise this would be the preferred option.
  * As part of the browse everything upload job, put the fileset id into the data object/s once the upload completes.
* TODO: How will these objects be cleaned up? message emitted / deleted as part of update?

Javascript-side implementation outline:
* using actioncable messages from the vue app should be pretty straightforward; see https://robots.thoughtbot.com/talking-to-actioncable-without-rails for a good example / guide.
* As mentioned above, subscribe to a channel for all fileset job-trackers on the resource in order to get new job-tracker ids as they are created, so as to subscribe to them.
* As those new job tracker channels are created, subscribe to them and add a fileset box.
  * on upload job success, channel should send a message with the fileset id so js can fetch title.
  * any action on characterization job success? I don't think so.
  * on create_derivatives success, js can fetch thumbnail from iiif server.

**Comment by hackmastera** [comment]
New idea for tracking job status. Each of the jobs we want to track (upload, characterize, derivatives) will kick off its own `JobStatusJob` which updates the status of the job, emits a message if the status has changed, and invokes itself again (after a short wait) if the job still exists. If the job it's tracking no longer exists, it deletes the `JobTracker` from the data store. Probably the `JobTracker` should delete its parent if it's the last child. This should be straightforward using the sidekiq-status gem: https://github.com/utgarda/sidekiq-status#retrieving-status

**Comment by hackmastera** [comment]
Also, sidekiq-status gives us exactly the sort of dashboard discussed in the initial ticket creation, so we could start using that very quickly / easily if desired. Although I'm not sure it sorts by queue which is unfortunate.

**Comment by hackmastera** [comment]
closing in favor of #638
